drivers: can: stm32fdcan

alexanderwachter opened the pull request:
STM32FDCAN implementation.
This PR implements a generic Bosch M_CAN driver that can be specialized for dedicated SoCs, such as FDCAN (STmicro), Microchip SAM, and NXP.
The specialization for FDCAN (STmicro) is included in this PR.
Tests are added to test the FD mode.

@legoabram has a prototype [prototype specialization](https://github.com/zephyrproject-rtos/zephyr/commit/192cbf55e515a07a5a54e067d48c84ccea00ae04) for SAM0

martinjaeger commented on drivers/can/Kconfig.stm32fd:
```diff
@@ -0,0 +1,40 @@
+	  Defines the maximum number of filters with extended ID (29-bit)
+	  that can be attached.
+
+config CAN_CKDIV
+	int "CKDIV register value"
+	range 0 15
+	default 0
+	help
+	  This value is written to the CKDIV register.
+	  The APB clock is divided according to this value before it is feed to
+	  CAN core. Note that the the divider affects all CAN controllers.
+	  The values of the register are multiplied by two and zero corresponds
+	  to one. The value six, for example results in a clock divided by 12.
```
I did not understand this explanation without looking into the manual, so I suggest below change and using `>> 2` in the code to get the value actually written to the register. This is also more consistent with other clock settings via Kconfig.

```suggestion
config CAN_STM32_CLOCK_DIVISOR
	int "CAN clock divisor"
	range 1 30
	default 1
	help
	  The APB clock is divided by this value (stored in CKDIV register)
	  before it is fed to the CAN core.
	  Note that the the divisor affects all CAN controllers.

	  Allowed values: 1 or 2 * n with n <= 15.
```

alexanderwachter replied:
IMO it is better if we don't allow invalid settings in the first place. Maybe I can come up with a better help text.

martinjaeger replied:
I think it's most important that Kconfig settings can be easily understood without knowing specific details of the underlying hardware.

It's very common in Zephyr to only mention the allowed values for clock divisors in the help text if it's not possible to check for correct values via Kconfig directly, see e.g. [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/clock_control/Kconfig.stm32g4).

alexanderwachter replied:
Changed. I added a compile time check to guarantee the n*2

martinjaeger replied:
Nice, the compile-time check is a good idea!

martinjaeger commented on include/drivers/can.h:
```diff
@@ -388,6 +389,46 @@ __subsystem struct can_driver_api {
+ *
+ * @retval Number of bytes
+ */
+
```
Nits: Remove blank line between comment and function. (also below)

martinjaeger commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -302,6 +302,54 @@
+				label = "CAN_1";
+			};
+
+			can2: can@40006800 {
+				compatible = "st,stm32-fdcan";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0x40006800 0x400>, <0x4000A800 0x350>;
+				reg-names = "m_can", "message_ram";
+				interrupts = <86 0>, <87 0>;
+				interrupt-names = "LINE_0", "LINE_1";
+				status = "disabled";
+				label = "CAN_2";
+			};
+
+			can3: can@40006C00 {
+				compatible = "st,stm32-fdcan";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0x40006C00 0x400>, <0x4000AC00 0x350>;
+				reg-names = "m_can", "message_ram";
+				interrupts = <88 0>, <89 0>;
+				interrupt-names = "LINE_0", "LINE_1";
+				status = "disabled";
+				label = "CAN_3";
+			};
```
This should go into `stm32g474.dtsi`, as the STM32G43X series only have one CAN.

alexanderwachter replied:
Thanks. Moved.

martinjaeger commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1535 @@
+#include <drivers/can.h>
+
+/*
+ * Register Masks are taken from the stm32cube library and extenden for
```
Nits:
```suggestion
 * Register Masks are taken from the stm32cube library and extended for
```

martinjaeger reviewed (commented):
Thanks @alexanderwachter for all your effort. I have tested the driver in a classic CAN network already and can confirm that it works very well.

Still need to go through the `can_mcan.c`, but here are some small comments already.

alexanderwachter committed:
drivers: can: Add Kconfig option for CANFD datalenght

Add Kconfig option to select appropriate can data field length

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

legoabram commented on drivers/can/can_mcan.c:
```diff
@@ -0,0 +1,845 @@
+			LOG_WRN("Dataphase bitrate error: %d", ret);
+		}
+	}
+#endif
```
This looks like it's still using some of the arbitration phase variables instead of data phase. As a side note, I noticed this since it was causing a crash in `can_calc_prescalar` since prescaler ended up with 0 resulting in a divide by zero exception.

legoabram commented on drivers/can/can_mcan.c:
```diff
@@ -0,0 +1,845 @@
+		}
+	}
+#ifdef CONFIG_CAN_FD_MODE
+	timing_data.sjw = cfg->sjw;
```
This looks like it's still using some of the arbitration phase variables instead of data phase. As a side note, I noticed this since it was causing a crash in `can_calc_prescalar` since prescaler ended up with 0 resulting in a divide by zero exception.

alexanderwachter replied:
Good catch. Thanks.

Nukersson commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1535 @@
+#define CAN_MCAN_IR_ARA_POS          (23U)
+#define CAN_MCAN_IR_ARA_MSK          (0x1UL << CAN_MCAN_IR_ARA_POS)
+#define CAN_MCAN_IR_ARA              CAN_MCAN_IR_ARA_MSK
+#else
```
```suggestion
#else  /* CONFIG_CAN_STM32FD */
```

Nukersson commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1535 @@
+#define CAN_MCAN_IR_ARA_POS          (29U)
+#define CAN_MCAN_IR_ARA_MSK          (0x1UL << CAN_MCAN_IR_ARA_POS)
+#define CAN_MCAN_IR_ARA              CAN_MCAN_IR_ARA_MSK
+#endif
```
```suggestion
#endif  /* CONFIG_CAN_STM32FD */
```

Nukersson commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1535 @@
+#define CAN_MCAN_IE_ARA_MSK         (0x1UL << CAN_MCAN_IE_ARA_POS)
+#define CAN_MCAN_IE_ARA             CAN_MCAN_IE_ARA_MSK
+
+#else
```
```suggestion
#else  /* CONFIG_CAN_STM32FD */
```

Nukersson commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1535 @@
+#define CAN_MCAN_IE_ARA_POS          (29U)
+#define CAN_MCAN_IE_ARA_MSK          (0x1UL << CAN_MCAN_IE_ARA_POS)
+#define CAN_MCAN_IE_ARA              CAN_MCAN_IE_ARA_MSK
+#endif
```
```suggestion
#endif  /* CONFIG_CAN_STM32FD */
```

Nukersson commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1535 @@
+#define CAN_MCAN_ILS_ARA_POS          (29U)
+#define CAN_MCAN_ILS_ARA_MSK          (0x1UL << CAN_MCAN_ILS_ARA_POS)
+#define CAN_MCAN_ILS_ARA              CAN_MCAN_IL_ARA_MSK
+#endif
```
```suggestion
#endif  /* CONFIG_CAN_STM32FD */
```

Nukersson commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1535 @@
+#define CAN_MCAN_RXGFC_LSE_POS       (24U)
+#define CAN_MCAN_RXGFC_LSE_MSK       (0xFUL << CAN_MCAN_RXGFC_LSE_POS)
+#define CAN_MCAN_RXGFC_LSE           CAN_MCAN_RXGFC_LSE_MSK
+#else
```
```suggestion
#else  /* CONFIG_CAN_STM32FD */
```

Nukersson commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1535 @@
+#define CAN_MCAN_XIDFC_LSS_POS      (16U)
+#define CAN_MCAN_XIDFC_LSS_MSK      (0x7FUL << CAN_MCAN_XIDFC_LSS_POS)
+#define CAN_MCAN_XIDFC_LSS          CAN_MCAN_XIDFC_LSS_MSK
+#endif
```
```suggestion
#endif  /* CONFIG_CAN_STM32FD */
```

Nukersson commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1535 @@
+	volatile uint32_t txefs;    /* Tx Event FIFO Status */
+	volatile uint32_t txefa;    /* Tx Event FIFO Acknowledge */
+};
+#else /* CONFIG_SOC_SERIES_STM32G4X */
```
Is it not same for H7 also?

alexanderwachter replied:
Yes, it is - updated comment.

KozhinovAlexander reviewed (changes requested):
Thank you very much for this PR. I would also test with it with my stm32h7 soon and go through ToDo's.

martinjaeger commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -302,6 +302,54 @@
 			label = "SPI_3";
 		};
 
+		can {
```
This leads to a warning during DTS generation:
```
board_name.dts.pre.tmp:281.7-324.5: Warning (simple_bus_reg): /soc/can: missing or empty reg/ranges property
```

alexanderwachter replied:
@mbolivar any suggestions on how we can improve the DT bindings?

martinjaeger replied:
Maybe the top-level node could be set to the general CANFD configuration register?
```suggestion
		can: can@40006500 {
```

erwango replied:
It possibly miss a reg = <0>;

alexanderwachter commented on boards/arm/nucleo_g474re/nucleo_g474re.dts:
```diff
@@ -130,3 +131,20 @@
+	bus-speed = <125000>;
+	sjw = <1>;
+	sample-point = <875>;
+	/*prop-seg = <0>;
+	phase-seg1 = <21>;
+	phase-seg2 = <3>;*/
```
I'll delete the manual timings. The manual timings for the other boards are mostly wrong. I'd suggest switching to the automatic timing parameter calculations for the other boards too.

alexanderwachter commented on drivers/can/can_mcan.c:
```diff
@@ -0,0 +1,845 @@
+	/* No memset because only aligned ptr are allowed */
+	for (uint32_t *ptr = (uint32_t *)msg_ram;
+	     ptr < (uint32_t *)msg_ram +
+		   sizeof(struct can_mcan_msg_sram) / sizeof(uint32_t);
```
ARRAY_SIZE can't be used because the elements are structs. It is not an array, but a struct of structs of arrays

alexanderwachter commented on drivers/can/can_mcan.c:
```diff
@@ -0,0 +1,845 @@
+	filter_nr = can_mcan_get_free_ext(msg_ram->ext_filt);
+
+	if (filter_nr == CAN_NO_FREE_FILTER) {
+		LOG_INF("No free ext filter left");
```
Not sure here if we should even lower to DBG. The user is responsible for checking the return value and printing an error.

martinjaeger commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -302,6 +302,42 @@
+				label = "CAN_1";
+			};
+
+			can2: can@40006800 {
```
`can2` is also not present in the G431, so it should go to `stm32g474.dtsi`.

alexanderwachter committed:
tests: driver: can: api: Enhanced tests to check dispatching

This commit enhances the CAN API test. In the send_receive tests,
we are using two filters and frames at the same time to check
if the frame gets dispatched to the correct filter.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
tests: driver: can: canfd: Add CAN-FD tests

This commit adds tests for CAN-FD frames.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter commented:
@henrikbrixandersen @erwango @FRASTM this PR is ready for review now.

KozhinovAlexander reviewed (changes requested):
I would prefer to add following files in separate commits:

1. device-tree: add fd-can to stm32g4 soc:
    dts/arm/st/g4/stm32g4.dtsi
    dts/arm/st/g4/stm32g474.dtsi

2. board: arm: nucleo_g474re: add fd-can
    boards/arm/nucleo_g474re/nucleo_g474re.dts
    boards/arm/nucleo_g474re/nucleo_g474re.yaml

The reason is the SoC and board does not have to do with driver development.

martinjaeger reviewed (changes requested):
Thanks for the updates. Finished reviewing all files except for `can_mcan.c`, which I hope to be able to review till end of the week as well. Here are some further comments already.

legoabram commented on drivers/can/can_mcan.c:
```diff
@@ -0,0 +1,867 @@
+		}
+	}
+#ifdef CONFIG_CAN_FD_MODE
+	timing_data.sjw = cfg->sjw;
```
This `sjw` should be updated to `sjw_data`. Also, these values appear to be assigned into the timing structs at the beginning of the function, so `timing.sjw = cfg->sjw;` and `timing_data.sjw = cfg->sjw;` could be removed.

alexanderwachter replied:
Moved the assignment just before the timing settings. Thanks.

alexanderwachter commented on drivers/can/can_mcan.h:
```diff
@@ -0,0 +1,232 @@
+	struct can_mcan_mm mm; /* Message marker */
+} __packed;
+
+#define can_mcan_FCE_DISABLE    0x0
+#define can_mcan_FCE_FIFO0      0x1
+#define can_mcan_FCE_FIFO1      0x2
+#define can_mcan_FCE_REJECT     0x3
+#define can_mcan_FCE_PRIO       0x4
+#define can_mcan_FCE_PRIO_FIFO0 0x5
+#define can_mcan_FCE_PRIO_FIFO1 0x7
+
+#define can_mcan_SFT_RANGE       0x0
+#define can_mcan_SFT_DUAL        0x1
+#define can_mcan_SFT_MASKED      0x2
+#define can_mcan_SFT_DISABLED    0x3
+
+struct can_mcan_std_filter {
+	volatile uint32_t id2  : 11; /* ID2 for dual or range, mask otherwise */
+	volatile uint32_t res  :  5;
+	volatile uint32_t id1  : 11;
+	volatile uint32_t sfce :  3; /* Filter config */
+	volatile uint32_t sft  :  2; /* Filter type */
+} __packed;
+
+#define can_mcan_EFT_RANGE_XIDAM 0x0
+#define can_mcan_EFT_DUAL        0x1
+#define can_mcan_EFT_MASKED      0x2
+#define can_mcan_EFT_RANGE       0x3
```
Yep, find and replace error :-(

alexanderwachter committed:
drivers: can: Implement stm32fd driver

This driver is the SoC specific implementation of the
Bosch M_CAN IP.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
dts: arm: stm32g4: add can bindings to stm32g4 SoCs

This commit adds the CAN nodes to the STM32g4 SoCs.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
board: nucleo_g474re: Add CAN support.

This commit adds CAN support to the nucleo_g474re board.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

martinjaeger commented on drivers/can/can_stm32fd.c:
```diff
@@ -0,0 +1,282 @@
+	LL_RCC_SetFDCANClockSource(LL_RCC_FDCAN_CLKSOURCE_PCLK1);
+	__HAL_RCC_FDCAN_CLK_ENABLE();
+
+	FDCAN_CONFIG->CKDIV = CONFIG_CAN_STM32_CLOCK_DIVISOR << 1;
```
Shouldn't this be right-shifted by 1?

alexanderwachter replied:
Yes.

alexanderwachter commented:
`#if CONFIG_CAN_STM32_CLOCK_DIVISOR != 1 && CONFIG_CAN_STM32_CLOCK_DIVISOR & 0x01` is better.
Will be pushed in the next time.

Nukersson commented:
Can you plese move following files from the commit aae95adb83272fedfdc69354112749ad9c8584fc to corresponding board f42451c16d617debf219b52d8fefab3a262d1940 and SoC e5e591c3b14ce5d6e73f546956f7d0001e562075 commits:

 boards/arm/nucleo_g474re/nucleo_g474re.dts
 dts/arm/st/g4/stm32g4.dtsi 

P.S.: `git reset HEAD~6` or `git rebase -i HEAD~7` in your case may help.

alexanderwachter commented:
Seems that something went wrong in the split. I'll fix that

Nukersson commented:
@alexanderwachter Compiling your example for `nucleo_g474re` board gives me following warning:

```
warning: CAN_MAX_FILTER (defined at drivers/can/Kconfig.stm32:15, drivers/can/Kconfig.mcux:12,
drivers/can/Kconfig.mcp2515:29, drivers/can/Kconfig.loopback:19) was assigned the value '5' but got
the value ''. Check these unsatisfied dependencies: ((CAN_STM32 && CAN) || (CAN_MCUX_FLEXCAN && CAN)
|| (CAN_MCP2515 && CAN) || (CAN_LOOPBACK && CAN)) (=n). See
http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_CAN_MAX_FILTER.html and/or look up
CAN_MAX_FILTER in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.
```

Do you see same warning with your setup?

Nukersson commented on drivers/can/can_stm32fd.c:
```diff
@@ -0,0 +1,282 @@
+
+#define DT_DRV_COMPAT st_stm32_fdcan
+
+int can_stm32fd_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	*rate = LL_RCC_GetFDCANClockFreq(LL_RCC_FDCAN_CLKSOURCE) /
+		CONFIG_CAN_STM32_CLOCK_DIVISOR;
+	return 0;
+}
```
```suggestion
int can_stm32fd_get_core_clock(const struct device *dev, uint32_t *rate)
{
	ARG_UNUSED(dev);
	*rate = LL_RCC_GetFDCANClockFreq(LL_RCC_FDCAN_CLKSOURCE);

	if (LL_RCC_PERIPH_FREQUENCY_NO == *rate) {
		LOG_ERR("Oscillator is not ready");
		return -EINVAL;
	}

	*rate /= CONFIG_CAN_STM32_CLOCK_DIVISOR;

	return 0;
}
```

The function `LL_RCC_GetFDCANClockFreq` has return values.

Nukersson commented on drivers/can/can_mcan_int.h:
```diff
@@ -0,0 +1,1560 @@
+	volatile uint32_t tscv;     /* Timestamp Counter Value */
+	volatile uint32_t tocc;     /* Timeout Counter Configuration */
+	volatile uint32_t tocv;     /* Timeout Counter Value */
+	uint32_t res1[4];           /* Reserved (4) */
```
Do we need `volatile` on all res too?

KozhinovAlexander reviewed (dismissed):
I have some more changes to add

alexanderwachter committed:
drivers: can: Implement Bosch M_CAN driver

Implementation of the Bosch M_CAN IP driver.
This driver is just the base for a specific SoC implementation.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter commented:
From a discussion I had wit @legoabram, we discovered that we will need an additional device-tree parameter to set the TDCO (Transceiver Delay Compensation Offset).
This parameter is property of the transceiver an therefor should go to the device-tree

lowlander commented:
@alexanderwachter  this driver is not for the "normal" CAN on for example the F407,  right ?

alexanderwachter commented:
Exactly. Its for the newer series with FDCAN (M_CAN), like the stm32g4, g0 and H7

alexanderwachter commented on drivers/can/can_mcan.c:
```diff
@@ -0,0 +1,867 @@
+			       const struct can_timing *timing_data)
+{
+	if (timing) {
+		__ASSERT_NO_MSG(timing->prop_seg == 0);
```
I think there is no prop seg in M_CAN. It is combined in the first segment.

alexanderwachter commented:
> @alexanderwachter Compiling your branch for `samples/drivers/can` example with `nucleo_g474re` board gives me following warning:
> 
> ```
> warning: CAN_MAX_FILTER (defined at drivers/can/Kconfig.stm32:15, drivers/can/Kconfig.mcux:12,
> drivers/can/Kconfig.mcp2515:29, drivers/can/Kconfig.loopback:19) was assigned the value '5' but got
> the value ''. Check these unsatisfied dependencies: ((CAN_STM32 && CAN) || (CAN_MCUX_FLEXCAN && CAN)
> || (CAN_MCP2515 && CAN) || (CAN_LOOPBACK && CAN)) (=n). See
> http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_CAN_MAX_FILTER.html and/or look up
> CAN_MAX_FILTER in the menuconfig/guiconfig interface. The Application Development Primer, Setting
> Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
> too.
> ```
> 
> Do you see same warning with your setup?

Yes. I think we should remove this setting in the sample and set the defaults high enough.

Nukersson commented:
> Exactly. Its for the newer series with FDCAN (M_CAN), like the stm32g4, g0 and H7
In my opinion it is to early to speak about h7. I've tested it on h723 and it needs some changes due to different registers structure. I'll prepare an driver update as soon as your development of this branch will be in master.

raul-klg commented:
Hi! Thanks for your work on this branch.

I think I found a problem running the `tests/drivers/can/canfd/` example. I hope this is the right place to discuss this, otherwise please, accept my apologies and let me know where else I should address this information.

I have tried it merged with a slightly more recent zephyr (2235c71cedd8d17f53947639ddc5d7469192e8f4) on a nucleo_g0b1re board. For being able to use the driver on the g0b1 board I created a dts overlay for it (it is taken from your stm32g4 proposal):
```dts
/ {

    chosen {
                zephyr,can-primary = &can1;
    };

    soc {
                can {
                        compatible = "bosch,m-can-base";
                        #address-cells = <1>;
                        #size-cells = <1>;
                        std-filter-elements = <28>;
                        ext-filter-elements = <8>;
                        rx-fifo0-elements = <3>;
                        rx-fifo1-elements = <3>;
                        rx-buffer-elements = <0>;
                        tx-buffer-elements = <3>;

                        can1: can@40006400 {
                                compatible = "st,stm32-fdcan";
                                #address-cells = <1>;
                                #size-cells = <0>;
                                reg = <0x40006400 0x400>, <0x4000A400 0x350>;
                                reg-names = "m_can", "message_ram";
                                interrupts = <21 0>, <22 0>;
                                interrupt-names = "LINE_0", "LINE_1";
                                status = "disabled";
                                label = "CAN_1";
                        };
                };

    };
};

&can1 {
        pinctrl-0 = <&fdcan1_rx_pa11 &fdcan1_tx_pa12>;
        bus-speed = <125000>;
        sjw = <1>;
        sample-point = <875>;
        bus-speed-data = <1000000>;
        sjw-data = <1>;
        sample-point-data = <875>;
        status = "okay";
};
```
I then build the `tests/drivers/can/canfd/` example and run it on the target with the following serial output:

```

*** Booting Zephyr OS build zephyr-v2.5.0-1509-gc0835a12fd30  ***
Running test suite canfd_driver
===================================================================
START - test_set_loopback
 PASS - test_set_loopback in 0.1 seconds
===================================================================
START - test_send_receive_std

    Assertion failed at WEST_TOPDIR/sample.git/canfd/src/main.c:117: check_msg: (msg1->id not equal to msg2->id)
ID does not match
 FAIL - test_send_receive_std in 0.13 seconds
===================================================================
START - test_send_receive_fd
```

Apparently it stalls on the `test_send_receive_fd` test and also there is an assertion on the previous `test_send_receive_std` This does not look to me like a successful test run.

For the sake of not cluttering the PR, I am omitting some debug information I could provide if it can be helpful to anyone.

legoabram commented:
Something I noticed: `CAN_MAX_DLC` is hard coded to 8 even if FD mode is enabled with a higher `CAN_MAX_DLEN`. We will likely need an `#if` tree to make it reflect the max data length. That said, I think changing this define will cause issues since many modules support both classic and CAN-FD, and they will need to use the correct max dlc when checking data depending on the requested mode.

On that same note, a number of the existing drivers depend on this value while iterating through data bytes. Though it will likely work, it's not semantically correct since DLC is no longer explicitly the number of bytes with the introduction of CAN-FD.

alexanderwachter commented:
@raul-klg do the api tests pass?
I'll rerun the tests on my g4 when time permits.

raul-klg commented:
@alexanderwachter  Thanks for your reply.
Maybe I am not running the test properly. When I wrote https://github.com/zephyrproject-rtos/zephyr/pull/31061#issuecomment-808191672 I didn't enable the CAN_LOOPBACK and I also didn't connect the nucleo-g0b1re CAN bus to any other device or even bus termination.
I have build now the tests/drivers/can/api application with the same board overlay as detailed in  https://github.com/zephyrproject-rtos/zephyr/pull/31061#issuecomment-808191672

I have run three cases:
1. default configuration. CONFIG_CAN_LOOPBACK is not set and CONFIG_CAN_STM32FD=y
**->Test failed.**
```
*** Booting Zephyr OS build zephyr-v2.5.0-1509-gc0835a12fd30  ***
Running test suite can_driver
===================================================================
START - test_set_loopback
 PASS - test_set_loopback in 0.1 seconds
===================================================================
START - test_send_and_forget
 PASS - test_send_and_forget in 0.1 seconds
===================================================================
START - test_filter_attach
 PASS - test_filter_attach in 0.1 seconds
===================================================================
START - test_receive_timeout
 PASS - test_receive_timeout in 0.101 seconds
===================================================================
START - test_send_callback

    Assertion failed at WEST_TOPDIR/scratch/can-api/src/main.c:609: test_send_callback: (ret not equal to 0)
Missing TX callback
 FAIL - test_send_callback in 0.113 seconds
===================================================================
START - test_send_receive_std
```
2. Both CAN loopback and stm32fd driver enabled. CONFIG_CAN_LOOPBACK=y and CONFIG_CAN_STM32FD=y
**->Test crashed**
```
I: Init of CAN_LOOPBACK done
*** Booting Zephyr OS build zephyr-v2.5.0-1509-gc0835a12fd30  ***
Running test suite can_driver
===================================================================
START - test_set_loopback
 PASS - test_set_loopback in 0.1 seconds
===================================================================
START - test_send_and_forget
ASSERTION FAIL [z_spin_lock_valid(l)] @ WEST_TOPDIR/zephyr/include/spinlock.h:129
        Recursive spinlock 0x20000d14
E: r0/a1:  0x00000004  r1/a2:  0x00000081  r2/a3:  0x080095cd
E: r3/a4:  0x08005d81 r12/ip:  0x00000002 r14/lr:  0x08005d89
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x0800769a
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: (nil) (unknown)
E: Halting system
```
3. Only CAN loopback is enabled.can32fd driver is not: CONFIG_CAN_LOOPBACK=y, CONFIG_CAN_STM32FD is not set
**->Test passed**
```
I: Init of CAN_LOOPBACK done
*** Booting Zephyr OS build zephyr-v2.5.0-1509-gc0835a12fd30  ***
Running test suite can_driver
===================================================================
START - test_set_loopback
 PASS - test_set_loopback in 0.1 seconds
===================================================================
START - test_send_and_forget
 PASS - test_send_and_forget in 0.1 seconds
===================================================================
START - test_filter_attach
 PASS - test_filter_attach in 0.1 seconds
===================================================================
START - test_receive_timeout
 PASS - test_receive_timeout in 0.101 seconds
===================================================================
START - test_send_callback
 PASS - test_send_callback in 0.1 seconds
===================================================================
START - test_send_receive_std
 PASS - test_send_receive_std in 0.1 seconds
===================================================================
START - test_send_invalid_dlc
E: DLC of 9 exceeds maximum (8)
 PASS - test_send_invalid_dlc in 0.3 seconds
===================================================================
START - test_send_receive_ext
 PASS - test_send_receive_ext in 0.1 seconds
===================================================================
START - test_send_receive_std_masked
 PASS - test_send_receive_std_masked in 0.1 seconds
===================================================================
START - test_send_receive_ext_masked
 PASS - test_send_receive_ext_masked in 0.2 seconds
===================================================================
START - test_send_receive_buffer
 PASS - test_send_receive_buffer in 0.2 seconds
===================================================================
START - test_send_receive_wrong_id
 PASS - test_send_receive_wrong_id in 0.101 seconds
===================================================================
Test suite can_driver succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Please, let me know if I am doing anything wrong and how else could I help.

Urist-McGit commented on drivers/can/can_mcan.c:
```diff
@@ -0,0 +1,867 @@
+
+		if (can->ir & CAN_MCAN_IR_ARA) {
+			can->ir  = CAN_MCAN_IR_ARA;
+			LOG_ERR("Acces to reserved address");
```
I found a small typo:
```suggestion
			LOG_ERR("Access to reserved address");
```

flaminggoat commented:
I have been testing this CAN implementation on a ATSAME51 using the prototype specialisations mentioned in the OP.
I found their is no register state change isr so I had to add the following to can_sam0.c
```c
static void can_sam0_register_state_change_isr(const struct device *dev,
						can_state_change_isr_t isr)
{
	const struct can_sam0_config *cfg = DEV_CFG(dev);
	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
	struct can_mcan_reg *can = mcan_cfg->can;
	struct can_mcan_data *mcan_data = &DEV_DATA(dev)->mcan_data;
	struct can_mcan_msg_sram *msg_ram = cfg->msg_sram;

	mcan_data->state_change_isr = isr;

	if (isr == NULL) {
		can->ie &= ~CAN_MCAN_IE_EP;
	} else {
		can->ie |= CAN_MCAN_IE_EP;
	}
}
```
I also have found that the bitrate does not seem to be working as expected with this device tree I am getting a bitrate of 370k:
```
&can0 {
	status = "okay";
	bus-speed = <1000000>;
	sjw = <1>;
	sjw-data = <1>;
	sample-point-data = <875>;
	bus-speed-data = <1000000>;
};
```

alexanderwachter commented:
@raul-klg the test run on my stm32g474 sucessfully.
`west build -b nucleo_g474re zephyr/tests/drivers/can/api` and `nucleo_g474re zephyr/tests/drivers/can/canfd`

martinjaeger commented on drivers/can/can_mcan.c:
```diff
@@ -0,0 +1,875 @@
+		(can->crel & CAN_MCAN_CREL_DAY) >> CAN_MCAN_CREL_DAY_POS);
+
+#ifndef CONFIG_CAN_STM32FD
+	can->sidfc = ((uint32_t)msg_ram->std_filt & CAN_MCAN_SIDFC_FLSSA_MSK) |
+		     (ARRAY_SIZE(msg_ram->std_filt) << CAN_MCAN_SIDFC_LSS_POS);
+	can->xidfc = ((uint32_t)msg_ram->ext_filt & CAN_MCAN_XIDFC_FLESA_MSK) |
+		     (ARRAY_SIZE(msg_ram->ext_filt) << CAN_MCAN_XIDFC_LSS_POS);
+	can->rxf0c = ((uint32_t)msg_ram->rx_fifo0 & CAN_MCAN_RXF0C_F0SA) |
+		     (ARRAY_SIZE(msg_ram->rx_fifo0) << CAN_MCAN_RXF0C_F0S_POS);
+	can->rxf1c = ((uint32_t)msg_ram->rx_fifo1 & CAN_MCAN_RXF1C_F1SA) |
+		     (ARRAY_SIZE(msg_ram->rx_fifo1) << CAN_MCAN_RXF1C_F1S_POS);
+	can->rxbc = ((uint32_t)msg_ram->rx_buffer & CAN_MCAN_RXBC_RBSA);
+	can->txefc = ((uint32_t)msg_ram->tx_event_fifo & CAN_MCAN_TXEFC_EFSA_MSK) |
+		     (ARRAY_SIZE(msg_ram->tx_event_fifo) <<
+		     CAN_MCAN_TXEFC_EFS_POS);
+	can->txbc = ((uint32_t)msg_ram->tx_fifo & CAN_MCAN_TXBC_TBSA) |
+		    (ARRAY_SIZE(msg_ram->tx_fifo) << CAN_MCAN_TXBC_TFQS_POS);
```
This doesn't look correct to me. The start addresses should be relative to the `msg_ram` and not as the absolute address in the MCU's RAM. In addition to that, it has to be left-shifted by 2 bits for most of the registers.

martinjaeger replied:
Never mind, I understand now that the addresses are actually absolute addresses, but they the message RAM must be located in the first 64k so that a 16-bit address is sufficient.

martinjaeger reviewed (commented):
Still finishing final `can_mcan.c` review, but here are some remarks already and one point that's not clear for me. Maybe you can comment already, @alexanderwachter?

martinjaeger reviewed (approved):
LGTM, only got above minor comments left.

I've been using this code in classic mode for a while already without any issues.

Thanks again for all the time you put into getting this done @alexanderwachter!

raul-klg commented:
> @raul-klg the test run on my stm32g474 sucessfully.
> `west build -b nucleo_g474re zephyr/tests/drivers/can/api` and `west build -b nucleo_g474re zephyr/tests/drivers/can/canfd`

Thanks for the verification (and for your additional work). I'll retry when I'm able to. Maybe I need to review how I used the branch.
